feat(plugin-dropzone-image): create plugin

shn-srl opened the pull request:
New Dropzone Image Plugin

Uses react-dnd and [react-resizable ](https://react-grid-layout.github.io/react-resizable/index.html)

The plugin uses the standard plugin approach, and has its definition in `packages/editor/src/plugins/dropzone-image/index.ts`
`static.tsx` renders the static view
`editor.tsx` renders the editing view

The main functionality of the drag drop and resizing lives in `packages/editor/src/plugins/dropzone-image/components/editor/EditorCanvas.tsx`

Other components will be found in `packages/editor/src/plugins/dropzone-image/components`

shn-srl committed:
MVP

elbotho commented on package.json:
```diff
@@ -56,5 +57,8 @@
     "node": ">=20"
+  },
+  "dependencies": {
+    "react-resizable": "^3.0.5"
```
this (and the types import above) should not be in the root `package.json`

elbotho commented on apps/web/src/serlo-editor-integration/create-plugins.tsx:
```diff
@@ -84,6 +85,12 @@ export function createPlugins({
+    {
+      type: EditorPluginType.DragDropBg,
+      plugin: createDragDropBgPlugin(),
+      visibleInSuggestions: true,
```
always good to do it now :) 
```suggestion
      visibleInSuggestions: !isProduction,
```

shn-srl replied:
Done :heavy_check_mark:

elbotho commented on apps/web/src/data/en/index.ts:
```diff
@@ -692,6 +692,13 @@ export const loggedInData = {
           emptyContentWarning: 'Boxes without content will not be displayed',
         },
+        dragDropBg: {
+          title: 'Drag & Drop Background',
```
@GregorZupan do you have a good plugin name suggestion?

elbotho commented on packages/editor/src/core/sub-document/editor.tsx:
```diff
@@ -87,6 +87,7 @@ export function SubDocumentEditor({ id, pluginProps }: SubDocumentProps) {
     if (!document.plugin) {
       // eslint-disable-next-line no-console
       console.warn('SubDocumentEditor -> Document is invalid')
+      console.log(document)
```
remove or add another `// eslint-disable-next-line no-console`?

shn-srl replied:
Done :heavy_check_mark:

elbotho commented on packages/editor/src/editor-integration/create-basic-plugins.tsx:
```diff
@@ -47,6 +47,12 @@ export function createBasicPlugins(props: Required<PluginsConfig>) {
+      type: EditorPluginType.DragDropBg,
+      plugin: createSpoilerPlugin({ allowedPlugins: allowedChildPlugins }),
+      visibleInSuggestions: true,
+      icon: <div>!!!</div>,
```
@GregorZupan do you want to build one?

elbotho commented on packages/editor/src/plugins/drag-drop-bg/components/answer-zone/AnswerZone.tsx:
since all other files in this repo follow kebab-case this plugin should as well.

elbotho commented on packages/editor/src/plugins/drag-drop-bg/components/answer-zone/AnswerZone.tsx:
```diff
@@ -0,0 +1,151 @@
+          width: dimensions.width,
+          height: dimensions.height,
+        }}
+        data-testid="box"
```
if this is meant for e2e: we use `data-qa="…id"` everywhere else. you can them target them with `$...id` in the e2e tests.

shn-srl replied:
Done :heavy_check_mark:

elbotho commented on packages/editor/src/plugins/drag-drop-bg/components/answer-zone/AnswerZone.tsx:
```diff
@@ -0,0 +1,151 @@
+        >
+          <div
+            ref={drag}
+            style={{ width: '100%', height: '100%', position: 'relative' }}
```
please use tailwind where feasible. you can get an introduction from any of the frontend people.

shn-srl replied:
Done :heavy_check_mark:

elbotho reviewed (changes requested):
Thanks for your work here!
Added some general comments how you could align the PR more the code style of other plugins. 

Additional TODOs:
- i18n (currently there are mostly hardcoded strings, some english, some german)
- using tailwind where possible would make this fit better into the codebase and we could avoid style duplication

If you want support or have question: just get in touch.

elbotho committed:
prettier

elbotho commented on packages/editor/src/plugins/drag-drop-bg/index.ts:
```diff
@@ -0,0 +1,71 @@
+  list,
+} from '../../plugin'
+
+export const answerZoneImageProps = child({ plugin: EditorPluginType.Image })
```
```suggestion
export const answerZoneImageState = child({ plugin: EditorPluginType.Image })
```

elbotho commented on packages/editor/src/plugins/drag-drop-bg/index.ts:
```diff
@@ -0,0 +1,71 @@
+  list,
+} from '../../plugin'
+
+export const answerZoneImageProps = child({ plugin: EditorPluginType.Image })
+export const answerZoneTextProps = child({ plugin: EditorPluginType.Text })
```
since they are not reused anywhere I'd prefer no extra vars for them (or at least no export)

elbotho replied:
This also goes for the other definitions here. 
In general: `Props` naming confuses me. We usually only use that for React Component Prop Types

shn-srl replied:
Done :heavy_check_mark:

elbotho commented on packages/editor/src/plugins/drag-drop-bg/index.ts:
```diff
@@ -0,0 +1,71 @@
+
+const dragDropBgState = object({
+  answerZones: list(answerZoneProps, 0),
+  backgroundType: string(''),
```
string is fine here, but I'd prefer an enum for the options so I can easily see what types are supported.

hejtful replied:
Added enums for backgroundType and backgroundShape, thanks for the suggestion 👍

elbotho commented on packages/editor/src/plugins/drag-drop-bg/index.ts:
```diff
@@ -0,0 +1,71 @@
+const dragDropBgState = object({
+  answerZones: list(answerZoneProps, 0),
+  backgroundType: string(''),
+  backgroundImage: child({ plugin: EditorPluginType.Image }),
```
Since there is a `backgroundType: 'text'` should this maybe be `optional()`?
And if there is only text/image as types could we maybe just remove `backgroundType` and use the image if it is defined?

elbotho replied:
something like:
``` ts
const dragDropBgState = object({
  answerZones: list(answerZoneProps, 0),
  backgroundImage: optional(child({ plugin: EditorPluginType.Image })),
  extraDraggableAnswers: list(wrongAnswerProps, 0),
})
```

shn-srl replied:
I had difficulties dealing with optional state properties in other places, the defined() check or trying to get the value of the property seemed tricky.

We could maybe pair on this tomorrow if it is important

elbotho replied:
> if it is important

Getting the state right is important imo (because it's quite annoying to change after there is content already created with an older version). But of course I don't know what the best state is :)
Let's see if we find a time for pairing.

In general it's a bit more code to work with optionals but it works quite well.
We have a lot of examples in the existing code.

hejtful replied:
Made optional, thanks for the suggestion 👍

elbotho commented on packages/editor/src/plugins/drag-drop-bg/static.tsx:
```diff
@@ -0,0 +1,86 @@
+          const dropZone = { ...rest, id: index.toString() }
+          return (
+            <BlankDropZone
+              key={`dustbin-${index}`}
```
would that work?
```suggestion
              key={id}
```

shn-srl committed:
Use tailwind instead of styles

shn-srl committed:
PR feedback changes

shn-srl committed:
Improve naming, make more concise, ESLint fixes

shn-srl committed:
Fix bug

shn-srl committed:
Improve handles design

shn-srl committed:
Fix clickability of settings button

shn-srl committed:
Add canvasShape property to state, and render choice in UI

shn-srl committed:
Fix default dropZone width/height

shn-srl committed:
Improve UI of possibleAnswers

shn-srl committed:
Improve behaviour of text answers in PossibleAnswers

shn-srl committed:
Fix dropping of text answers

shn-srl committed:
Setup for allowing multiple correct answers for a dropZone

shn-srl committed:
Use context

shn-srl committed:
Remove unused code

shn-srl committed:
Fix possibleAnswers in static view

shn-srl committed:
Move Ablagezone button to left side of toolbar

shn-srl committed:
Make settings button bottom right and visible only on hover

shn-srl committed:
Add border to text answers

shn-srl committed:
Improve settings modal

shn-srl committed:
Improve dragging UI

shn-srl committed:
Documentation and cleanup

shn-srl committed:
Show name in static view dropzones

shn-srl committed:
small code improvement

shn-srl committed:
Refactor

shn-srl committed:
Fix package installation

shn-srl committed:
Refactor

shn-srl committed:
Document

shn-srl committed:
Improvements, make better use of context

shn-srl committed:
Improvements

Add visibility toggle

Improve settings modal

CodingDive commented on apps/web/src/serlo-editor-integration/create-renderers.tsx:
```diff
@@ -77,6 +78,14 @@ const SpoilerSerloStaticRenderer = dynamic<
+  EditorDragDropBgDocument & { openOverwrite?: boolean; onOpen?: () => void }
+>(() =>
+  import('@editor/plugins/drag-drop-bg/static').then(
+    (mod) => mod.DragDropBgStaticRenderer
```
Nit: `module` makes this more clear

shn-srl replied:
Do you mean the naming of the argument, so renaming mod to module ? Or something else?

CodingDive replied:
Yes, just mod => module. I dislike abbreviations in general

elbotho replied:
Out of scope here imo. We use this syntax for all dynamic imports in the codebase. 
@CodingDive if it's important for you feel free to open a follow up where you change all occurences.

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/README.md:
```diff
@@ -0,0 +1,96 @@
+
+- `newWrongAnswer`: The current wrong answer being edited.
+
+### ModalWithCloseButton
```
Already existed and is in use throughout the code, so not sure if we need to document it here.

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZone.tsx:
```diff
@@ -0,0 +1,188 @@
+import { useDrag } from 'react-dnd'
+import { ResizableBox, ResizableBoxProps } from 'react-resizable'
+
+// TODO: Correctly import styles
```
This should be the correct style import already 😊 
Just delete the todo comment

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZone.tsx:
```diff
@@ -0,0 +1,188 @@
+import { FaIcon } from '@/components/fa-icon'
+
+/**
+ * AnswerZone component props
```
Delete ubiquitous comment

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZone.tsx:
```diff
@@ -0,0 +1,188 @@
+}
+
+/**
+ * AnswerZone component
```
"AnswerZone component" comment not needed imo

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZone.tsx:
```diff
@@ -0,0 +1,188 @@
+    const answerImageUrl = getAnswerZoneImageSrc(answer.image.get())
+    const answerText = getAnswerZoneText(answer.text.get())
+
+    const hasImage = !!answerImageUrl
```
I personally prefer `Boolean(answerImageUrl)`, but this may not be the code style of the rest of the repo 🤔

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZone.tsx:
```diff
@@ -0,0 +1,188 @@
+        onResize={handleResize}
+        onResizeStop={handleResizeStop}
+        resizeHandles={['nw', 'ne', 'sw', 'se']}
+        style={{
```
Can we pass `className` here instead of style?

shn-srl replied:
Done

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZone.tsx:
```diff
@@ -0,0 +1,188 @@
+            <>
+              <button
+                className="absolute right-2 top-1 z-20 rounded bg-orange-100 p-1"
+                style={{ fontSize: '.5rem' }}
```
Try if `text-xs` gives you similar size, if not I'd prefer the [] syntax of tailwind

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZone.tsx:
```diff
@@ -0,0 +1,188 @@
+              </button>
+              <button
+                className="absolute bottom-1 right-2 z-20 hidden rounded bg-orange-100 p-1 group-hover:block"
+                style={{ fontSize: '.5rem' }}
```
Same here, see comment above

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZoneSettingsForm.tsx:
```diff
@@ -0,0 +1,113 @@
+/* eslint-disable @typescript-eslint/unbound-method */
```
Is this still needed? If so, I prefer having `eslint-disable-next-line` in the middle of the code!

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZoneSettingsForm.tsx:
```diff
@@ -0,0 +1,113 @@
+import { faClone, faTrashCan } from '@fortawesome/free-regular-svg-icons'
+import React from 'react'
+
+import type { answerZoneType } from '../../types.js'
```
.js

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/AnswerZoneSettingsForm.tsx:
```diff
@@ -0,0 +1,113 @@
+}
+
+/**
+ * AnswerZoneSettingsForm component
```
Remove this line, rest of the comment other than jsdoc style params/return type looks great!

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/NewAnswerZoneFlow.tsx:
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
```
Prefer eslint-disable-next-line

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/NewAnswerZoneFlow.tsx:
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
```
Same here

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/AnswerZone/NewAnswerZoneFlow.tsx:
```diff
@@ -0,0 +1,85 @@
+}
+
+/**
+ * NewAnswerZoneFlow component
```
Remove this and the jsdoc style type/return type

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/editor/EditorCanvas.tsx:
```diff
@@ -0,0 +1,188 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
```
Prefer eslint-disable-next-line and one line below too

CodingDive replied:
(Or getting rid of it entirely)

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/editor/EditorCanvas.tsx:
```diff
@@ -0,0 +1,188 @@
+
+/**
+ * Get the image source URL for an answer zone.
+ * @param {string} answerZoneImageId - The ID of the answer zone image.
```
No need for jsdoc param and return string since we have TypeScript. Could make the return type explicit with e.g `type ImageSourceUrl = string`

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/editor/EditorCanvas.tsx:
```diff
@@ -0,0 +1,188 @@
+}
+
+/**
+ * Get the text content for an answer zone.
```
Remove whole comment. Code is very clear

shn-srl replied:
Done :heavy_check_mark:

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/editor/EditorCanvas.tsx:
```diff
@@ -0,0 +1,188 @@
+ * This component represents the canvas area where answer zones and possible answers are managed and displayed.
+ * It supports adding, editing, and deleting answer zones as well as managing possible answers.
+ *
+ * @param {DragDropBgProps} props - The properties for the EditorCanvas component.
```
Remove param (it's out of date anyway)

shn-srl replied:
Done ✔️

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/components/shared/WrongAnswerFlow.tsx:
```diff
@@ -0,0 +1,61 @@
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
```
Prefer eslint-disable-next-line or a fix

CodingDive commented on packages/editor/src/plugins/drag-drop-bg/editor.tsx:
```diff
@@ -0,0 +1,135 @@
+import { FaIcon } from '@/components/fa-icon'
+
+/**
+ * DragDropBgEditor component
```
Remove this line of the comment and param

shn-srl replied:
Done ✔️

CodingDive reviewed (commented):
Amazing work! @shn-srl 👏 🥳 
I left a few style comments, but nothing too critical. I18n is missing. Once you work some more on the UI, I'm happy to give you a round of UI/UX feedback.

Super excited for this plugin!

shn-srl committed:
Address PR feedback, remove old comments, cleanup

shn-srl committed:
Update readme

shn-srl committed:
Update comment

elbotho commented on apps/web/src/serlo-editor-integration/create-plugins.tsx:
```diff
@@ -84,6 +85,12 @@ export function createPlugins({
+      type: EditorPluginType.DragDropBg,
+      plugin: createDragDropBgPlugin(),
+      visibleInSuggestions: !isProduction,
+      icon: <div>!!!!</div>,
```
Is this being worked on?

elbotho commented on packages/editor/package.json:
```diff
@@ -44,8 +44,10 @@
   },
   "dependencies": {
     "@serlo/katex-styles": "1.0.1",
+    "@types/react-resizable": "^3.0.7",
```
types should be dev dependencies

shn-srl replied:
Done :heavy_check_mark:

elbotho commented on package.json:
```diff
@@ -45,9 +45,13 @@
     "start": "yarn workspace @serlo/frontend start",
     "test": "yarn workspace @serlo/frontend test"
   },
+  "dependencies": {
+    "react-resizable": "^3.0.5"
+  },
   "devDependencies": {
     "@serlo/eslint-config": "workspace:*",
     "@serlo/typescript-config": "workspace:*",
+    "@types/react-resizable": "^3.0.7",
```
please remove them here, they should not be in the monorepo root `package.json`

shn-srl replied:
Done :heavy_check_mark:

elbotho commented on packages/editor/src/plugins/drag-drop-bg/components/static/BlankDropZone.tsx:
```diff
@@ -0,0 +1,76 @@
+/* eslint-disable no-duplicate-imports */
```
```suggestion
```

shn-srl replied:
Done :heavy_check_mark:

shn-srl committed:
Fix dependencies and remove obsolete comment

shn-srl committed:
Fix static view class

shn-srl committed:
fix eslint error

shn-srl committed:
Use slate for text rendering

hejtful committed:
chore(yarn): push arm64 specific files

shn-srl committed:
Merge branch '240425/feat/drag-drop-plugin' of github.com:serlo/frontend into 240425/feat/drag-drop-plugin

shn-srl committed:
Rename all files, use kebab case :(

shn-srl committed:
Fix lint errors

shn-srl committed:
Remove readme

shn-srl committed:
Add icon

hejtful committed:
refactor(drag-drop-bg): make backgroundImage state prop optional

+ extract some components from the main editor component and replace abbreviations with complete words

shn-srl committed:
Add copy/paste feature for anserZones

shn-srl committed:
Add functionality to delete dropzone using backspace/delete key

shn-srl committed:
Improve display of items in answerZone

shn-srl committed:
improve styling

elbotho commented on packages/editor/src/plugins/drag-drop-bg/components/answer-zone/answer-zone.tsx:
```diff
@@ -120,6 +122,7 @@ export const AnswerZone = (props: AnswerZoneProps) => {
     <div
       ref={dragPreview}
       className="absolute flex cursor-move items-center justify-center rounded bg-white"
+      onClick={() => onClick && onClick(answerZone.id.get())}
```
I learned from @hejtful that this also a thing that works :) 
(no change request, just a note)
```suggestion
      onClick={() => onClick?.(answerZone.id.get())}
```

shn-srl committed:
Fix settings modal layout

shn-srl committed:
Fix duplicate button

hejtful committed:
refactor(drag-drop-bg): use enums, capitalize interface, remove unnecessary type

hejtful committed:
feat(drag-drop-bg): shuffle possible answers in static view

shn-srl committed:
Change visibility option UI and behaviour

shn-srl committed:
Use enum for Dropzone visibility

shn-srl committed:
Allow more than one dropped answer in static view drop zones

shn-srl committed:
Update check answers logic for multiple answers

hejtful committed:
feat(drag-drop-bg): edit and remove answers

shn-srl committed:
chore(drag-drop-bg) improve folder structure

shn-srl committed:
refactor(drag-drop-bg): Move functions to hook

shn-srl committed:
feat(drag-drop-bg): Add possibility to create wrong answers

hejtful committed:
feat(drag-drop-bg): drag answers from drop zone

hejtful committed:
merge with origin

hejtful committed:
fix(drag-drop-bg): add ids to wrong answers

hejtful committed:
fix(drag-drop-bg): only shuffle possible answers when they change

shn-srl committed:
feat(drag-drop-bg): Fix checkAnswers function

shn-srl committed:
refactor(drag-drop-bg): Use shared functions, more concise

shn-srl committed:
refactor(drag-drop-bg): Use one component for creation of AnswerZones or Wrong Answers

hejtful committed:
fix(drag-drop-bg): droppableBlankId consistently optional

shn-srl committed:
refactor(drag-drop-bg): Create AnswerRenderer component

shn-srl committed:
refactor(drag-drop-bg): Respect the enums

shn-srl committed:
refactor(drag-drop-bg): Remove unused properties from AnswerZone, visible and lockedAspectRatio

shn-srl committed:
fix(drag-drop-bg): Remove unused var

hejtful committed:
feat(drag-drop-bg): preview switch in toolbar

shn-srl committed:
feat(drag-drop-plugin): Write first test for plugin

shn-srl committed:
chore(drag-drop-plugin): remove unused vars

hejtful committed:
fix(drag-drop-bg): focus text plugin when adding text answer

+ split modal rendering from EditorCanvas

hejtful committed:
Merge branch '240425/feat/drag-drop-plugin' of https://github.com/serlo/frontend into 240425/feat/drag-drop-plugin

hejtful committed:
fix(drag-drop-bg): show partial visibility in static view

+ use enums all around

kulla commented on packages/editor/src/editor-integration/create-basic-plugins.tsx:
```diff
@@ -47,6 +47,12 @@ export function createBasicPlugins(props: Required<PluginsConfig>) {
+      type: EditorPluginType.DragDropBg,
+      plugin: createSpoilerPlugin(),
+      visibleInSuggestions: true,
+      icon: <div>!!!</div>,
```
Is the icon missing here?

CodingDive replied:
It is, but Yannick already designed one 🎉

shn-srl replied:
Included now :heavy_check_mark:

hejtful committed:
fix(drag-drop-bg): tooltip preview button visual bug

hejtful committed:
fix(drag-drop-bg): answer zone initial dimensions

shn-srl committed:
Basic correct WrongAnswers implementation

shn-srl committed:
Improve style

shn-srl committed:
fix(drag-drop-bg): Include icon in create-basic-plugins

shn-srl committed:
feat(drag-drop-bg): Improve UI for bg shape/type selection, include icons

shn-srl committed:
include icons

hejtful committed:
fix(drag-drop-bg): answer type icons and i18n strings

hejtful committed:
refactor(editor): extract toolbar preview button component

shn-srl committed:
feat(drag-drop-bg): Extend test

shn-srl committed:
Merge branch '240425/feat/drag-drop-plugin' of github.com:serlo/frontend into 240425/feat/drag-drop-plugin

hejtful committed:
feat(drag-drop-bg): background image settings

hejtful committed:
Merge branch '240425/feat/drag-drop-plugin' of https://github.com/serlo/frontend into 240425/feat/drag-drop-plugin

hejtful committed:
fix(drag-drop-bg): render background image settings modal conditionally

hejtful committed:
fix(drag-drop-bg): resize answerZone from settings

hejtful committed:
fix(drag-drop-bg): separate drag types and assign them to constants

shn-srl committed:
feat(drag-drop-bg): Update tests

shn-srl committed:
feat(drag-drop-bg): Background Shape Select option only shown for blank background

shn-srl committed:
refactor(drag-drop-bg): Change file structure

hejtful committed:
refactor(dropzone-image): rename drag-drop-bg to dropzone-image in the code

shn-srl committed:
fix(drag-drop-bg): remove margin on editor canvas

hejtful committed:
fix(dropzone-image): remove single answer hover background

hejtful committed:
refactor(dropzone-image): extract AnswerZoneAnswer component + answer-zone utils

shn-srl committed:
feat(drag-drop-bg): Resize canvas to background image

hejtful committed:
fix(dropzone-image): answer zone max dimensions

shn-srl committed:
Merge branch '240425/feat/drag-drop-plugin' of github.com:serlo/frontend into 240425/feat/drag-drop-plugin

shn-srl committed:
fix(drag-drop-bg): fix merge bug

shn-srl committed:
fix(drag-drop-bg): Set height and width on state when shape of canvas selected

shn-srl committed:
fix(drag-drop-bg): default height and width for canvas dimensions in state

hejtful committed:
fix(dropzone-image): transparent answer background on blank canvas

hejtful committed:
Merge branch '240425/feat/drag-drop-plugin' of https://github.com/serlo/frontend into 240425/feat/drag-drop-plugin

hejtful committed:
fix(dropzone-image): transparent answer zone background when visibility partial

hejtful committed:
fix(dropzone-image): extract possible and incorrect answers components

+ fix add incorrect answer button alignment + general cleanup

hejtful committed:
fix(dropzone-image): add answer wrong text

hejtful committed:
fix(dropzone-image): check answers button style

hejtful committed:
fix(dropzone-image): rename assets folder

shn-srl committed:
fix(drag-drop-bg): remove unusable resize handles

hejtful committed:
fix(dropzone-image): only show check answers button when all dropzones filled

hejtful committed:
fix(dropzone-image): thicker correct/incorrect answer zone border

+ cleaner and more complete border styling logic

hejtful committed:
fix(dropzone-image): thicker correct/incorrect answer border + spacing

shn-srl committed:
refactor(drag-drop-bg): Use editorStrings

shn-srl committed:
fix(drag-drop-bg): Only English strings

shn-srl committed:
fix(drag-drop-bg): Improve UX of checking answers

shn-srl committed:
feat(drag-drop-bg): improve UI, nicer borders

hejtful committed:
fix(dropzone-image): answer sizing in static and editor

hejtful committed:
merge staging into branch

hejtful committed:
fix(dropzone-image): remove unnecessary styling prop

hejtful committed:
fix(dropzone-image): adjust to new modal lib

kulla commented on packages/editor/src/plugins/dropzone-image/index.ts:
```diff
@@ -0,0 +1,70 @@
+  optional,
+} from '../../plugin'
+
+export const positionData = object({
+  top: number(0),
+  left: number(0),
+})
```
Is this relative or absolute positioning? I would argue for relative positioning...

shn-srl replied:
This is for the functionality of the canvas.

shn-srl replied:
I'm not sure why this is an important detail can you please explain?

The container is relative and the answerZones are absolute

shn-srl replied:
`packages/editor/src/plugins/dropzone-image/components/answer-zone/answer-zone.tsx:76`

shn-srl replied:
`packages/editor/src/plugins/dropzone-image/components/editor/editor-canvas.tsx:117`

kulla replied:
> I'm not sure why this is an important detail can you please explain?

Let's say you upload an image in 800x600 dimension. However due to a smaller screen the image is resized and shown in 400x300. Then the absolute positions of the drag & drop zones are messed up. This is relevant for showing the plugin in mobile devices.

The easiest is to store relative positions (top: 0-100% and left: 0-100%) and to position the drag & drop zones based on the current size of the image.

kulla replied:
Currently I see two things:

1. In the edit mode the image scales and thus the drop zones aren't defined properly when the screen changes
2. The static view is not mobile ready

See 

https://github.com/serlo/frontend/assets/1327215/b6d8b4a8-5970-4b05-91a9-ec2ddd8b791f

shn-srl replied:
Edit mode is desktop only, working on the fix for static view now

shn-srl replied:
Static view is mobile ready now.

I would call the screen changing size, and responsiveness to that, out of scope for this iteration

elbotho replied:
> I would call the screen changing size, and responsiveness to that, out of scope for this iteration

Meaning that we would do it later or never?

kulla commented on packages/editor/src/plugins/dropzone-image/index.ts:
```diff
@@ -0,0 +1,70 @@
+  left: number(0),
+})
+
+export const layoutData = object({
+  width: number(200),
+  height: number(70),
+})
```
How are those numbers defined currently? Size of original image? What happens on smaller displays when the image resizes...

shn-srl replied:
Good point, will try to make a change so the static view looks acceptable on mobile

kulla replied:
> Good point, will try to make a change so the static view looks acceptable on mobile

Nah, IMHO a good user experience on mobile for the static view is essential before merging this PR...

shn-srl replied:
Thats what I meant, I will make a change to make the mobile look on static acceptable. Mostly finished this change, just need to fix one more bug. Will push asap

shn-srl replied:
Pushed a fix and the static view works in mobile now. I think we would need to get this on staging to test more properly with mobile devices

shn-srl replied:
Going to mark this as resolved

hejtful committed:
fix(dropzone-image): answer checking logic

hejtful committed:
fix(dropzone-image): clean and comment answer checking logic

shn-srl committed:
feat(drag-drop-bg): Improve UX of static view on mobile

shn-srl committed:
Merge branch '240425/feat/drag-drop-plugin' of github.com:serlo/frontend into 240425/feat/drag-drop-plugin

hejtful committed:
feat(dropzone-image): answer zones feedback

hejtful committed:
refactor(dropzone-image): extract StaticCanvas component

hejtful committed:
refactor(dropzone-image): extract FeedbackButton component

hejtful committed:
refactor(dropzone-image): extract draggable area answer filtering out of jsx

elbotho commented:
I can definitely find some smaller layout inconsistencies. They could be related to the preview, but I'm not sure.
Examples:
- In the editor two dropzones that have a gap between them are touching in the static view
- the image is scaled differently in the static view causing the small dropbox to not start at the text 

<img width="1382" alt="image" src="https://github.com/serlo/frontend/assets/1258870/856f27cb-faa4-43a2-ae59-cce729857428">

elbotho commented:
Not a big issue for staging but for before this goes into production:
<img width="435" alt="image" src="https://github.com/serlo/frontend/assets/1258870/969021d1-5ecc-45cc-b7d3-84698fe01cd7">
We should not allow some of those options here, diabling them now is easy. Later is always a bit of pain :)
Imo this should be inline only (no headings, no lists) and also no links. Even italic and bold are arguable.

elbotho commented on apps/web/src/serlo-editor-integration/create-plugins.tsx:
```diff
@@ -84,6 +86,12 @@ export function createPlugins({
+      plugin: createDropzoneImagePlugin(),
+      visibleInSuggestions: !isProduction,
+      icon: <IconDropzones />,
+    },
```
I'd prefer to not load this at all in production (see how it's done for the audio plugin)

elbotho commented on e2e-tests/tests/460-drag-drop-plugin.ts:
```diff
@@ -0,0 +1,129 @@
+
+  I.click('$plugin-image-src')
+  I.type(
+    'https://upload.wikimedia.org/wikipedia/en/a/aa/Bart_Simpson_200px.png'
```
should we maybe use a serlo hosted image here?

elbotho commented on e2e-tests/tests/460-drag-drop-plugin.ts:
```diff
@@ -0,0 +1,129 @@
+
+    I.click('$plugin-image-src')
+    I.type(
+      'https://upload.wikimedia.org/wikipedia/en/a/aa/Bart_Simpson_200px.png'
```
see above

elbotho commented on packages/editor/src/editor-integration/create-basic-plugins.tsx:
```diff
@@ -47,6 +49,12 @@ export function createBasicPlugins(props: Required<PluginsConfig>) {
       visibleInSuggestions: true,
       icon: <IconSpoiler />,
     },
+    {
+      type: EditorPluginType.DropzoneImage,
+      plugin: createDropzoneImagePlugin(),
+      visibleInSuggestions: true,
+      icon: <IconDropzones />,
+    },
```
This would make the plugin immediately public in the integrations on the next update. Not sure this is intended.

elbotho commented on packages/eslint-config/editor.js:
```diff
@@ -64,7 +64,10 @@ module.exports = {
-    '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
+    '@typescript-eslint/no-unused-vars': [
+      'error',
+      { argsIgnorePattern: '^_', ignoreRestSiblings: true },
```
could you say something about the choice?

shn-srl replied:
@hejtful I'm not sure if my editor did this and I committed, or if it was a conscious choice by you?

hejtful replied:
I made this choice. @shn-srl had to add an eslint-ignore comment for [these lines](https://github.com/serlo/frontend/pull/3805/files#diff-bc35fb3fc9546511b24e7c499e87ec4db1eb7be43788c98390327386aefb62a4R53-R54). I thought that rest syntax is a normal part of modern JS, and that unused vars shouldn't stop us from using it.

It's worth mentioning that the changes in the code made the var in those lines used, so there is no need for this change in the config any more. Still, I'd keep this exception, so that we can freely use the rest syntax in the future. I am open to discussing it, of course :D

elbotho replied:
Thank you for the explanation. So just I understand: You want to use `...rest` to exclude fields from an object so you deliberately define the unused variables?

hejtful replied:
Yes.

But, "to exclude fields from an object" is more like "to create a new object without certain fields".

elbotho replied:
thank you

elbotho commented on packages/editor/src/plugins/dropzone-image/index.ts:
```diff
@@ -0,0 +1,70 @@
+  answerZones: list(answerZoneState, 0),
+  canvasShape: string(BackgroundShape.Unset),
+  canvasDimensions: object({
+    height: number(786),
+    width: number(786),
```
This feedback might be to late, but like Kulla I'd prefer a relative / percentage based approach for positioning and sizing.
E.g. `786px` as the base width could change and is already different in some integration I think.
(same then goes for `layoutData`)

shn-srl replied:
Yes, we will explore this for the 2nd iteration, for now we are using the scaling solution, which works sufficiently for small screens

LarsTheGlidingSquirrel replied:
The available space for the editor in the upcoming integrations will definitely vary. Sometimes it will be smaller than on serlo.org because the editor will be shown in an iframe on another page.
I guess this speaks for percentage positioning and sizing. But I am fine with this solution for now until we go live.

elbotho commented on packages/editor/src/plugins/dropzone-image/index.ts:
```diff
@@ -0,0 +1,70 @@
+export const defaultConfig: DropzoneImageConfig = {}
+
+export function createDropzoneImagePlugin(): EditorPlugin<DropzoneImagePluginState> {
+  // config = defaultConfig
```
```suggestion
```

elbotho commented on packages/editor/src/plugins/dropzone-image/components/static/static-canvas.tsx:
```diff
@@ -0,0 +1,70 @@
+  ) => void
+}
+
+export function StaticCanvas(props: StaticCanvasProps) {
```
I was fearing a real html-canvas for a second, phew 😄

elbotho commented on packages/editor/src/plugins/dropzone-image/components/answer-zone/new-answer-flow.tsx:
```diff
@@ -0,0 +1,91 @@
+      />
+    </div>
+  ) : (
+    <>Error </>
```
could this happen?

hejtful replied:
Good point, I didn't even notice this. @shn-srl , could you please explain this case shortly?

shn-srl replied:
Removed :+1:

elbotho reviewed (changes requested):
Thank you for working in my first feedbacks!
I requested some changes that should be solved before merging to staging so that we can deploy staging at any time and don't block it.
The rest should not stop you from merging and testing in staging.

I trust that you tested the changes in the other plugins (mostly exercise type toolbars and blanks exercise) and they still work as expected :)

hejtful committed:
fix(dropzone-image): hide the plugin in production and integrations

hejtful committed:
chore(dropzone-image): remove unnecessary comment

hejtful committed:
test(dropzone-image): change dummy image to Serlo-hosted one

hejtful committed:
test(dropzone-image): fix test name

shn-srl committed:
refactor(dropzone-image): remove old debug code

shn-srl committed:
Merge branch '240425/feat/drag-drop-plugin' of github.com:serlo/frontend into 240425/feat/drag-drop-plugin

shn-srl committed:
fix(dropzone-image-plugin): AnswerZone height

shn-srl commented:
> Thank you for working in my first feedbacks! I requested some changes that should be solved before merging to staging so that we can deploy staging at any time and don't block it. The rest should not stop you from merging and testing in staging.
> 
> I trust that you tested the changes in the other plugins (mostly exercise type toolbars and blanks exercise) and they still work as expected :)



> I can definitely find some smaller layout inconsistencies. They could be related to the preview, but I'm not sure. Examples:
> 
>     * In the editor two dropzones that have a gap between them are touching in the static view
> 
>     * the image is scaled differently in the static view causing the small dropbox to not start at the text
> 
> 
> <img alt="image" width="1382" src="https://private-user-images.githubusercontent.com/1258870/339269525-856f27cb-faa4-43a2-ae59-cce729857428.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyNzA0NjAsIm5iZiI6MTcxODI3MDE2MCwicGF0aCI6Ii8xMjU4ODcwLzMzOTI2OTUyNS04NTZmMjdjYi1mYWE0LTQzYTItYWU1OS1jY2U3Mjk4NTc0MjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTNUMDkxNjAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2JkNTRjNzk2N2I4NDc5MzJiNDBlN2JjMDhmNjkzMWZlMDVkZWEzYWRhZTI5NGFjMDJjY2Q2NmNkMjBlMzMyYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5zxEJ-cB4c42ke2Qh-ndYiXbTi7AzLIGDlxiyH7pH-g">

Great eye, thanks ! 

Fixed 

![image](https://github.com/serlo/frontend/assets/161819227/6e1038b9-28b5-4d7d-9e27-7cab01b5bfa1)
![image](https://github.com/serlo/frontend/assets/161819227/f4cb76d1-740b-49da-bf48-a6dd29c0a194)

shn-srl committed:
fix(dropzone-image-plugin): Fix layout sync of editor and static

hejtful commented:
>Not a big issue for staging but for before this goes into production:
image
We should not allow some of those options here, diabling them now is easy. Later is always a bit of pain :)
Imo this should be inline only (no headings, no lists) and also no links. Even italic and bold are arguable.

@elbotho Discussed this with Gregor, will leave formula and code, bold and italic, and color.

hejtful committed:
test(dropzone-image): fix affected tests

hejtful committed:
Merge branch '240425/feat/drag-drop-plugin' of https://github.com/serlo/frontend into 240425/feat/drag-drop-plugin

hejtful committed:
fix(dropzone-image): restrict available text formatting options

elbotho reviewed (approved):
let's try it in staging :)
